File: bindings/node/samples/pubsub_recv_sample.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const zlink = require('@zlink-systems/zlink');
const { tcpEndpoint } = require('./sample_support');

async function main() {
// --8<-- [start:doc]
  const endpoint = await tcpEndpoint();
  const ctx = zlink.createContext();
  const pub = zlink.createPubSocket(ctx);
  const sub = zlink.createSubSocket(ctx);

  try {
    const pubMonitor = pub.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    const subMonitor = sub.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
    try {
      pub.bind(endpoint);
      sub.connect(endpoint);
      pubMonitor.recv();
      subMonitor.recv();
    } finally {
      pubMonitor.close();
      subMonitor.close();
    }

    const topic = 'prices';
    const sent = '101.25';
    sub.setSubscription(topic);
    const deadline = Date.now() + 5000;
    const received = new zlink.TopicMessage();
    let hasReceived = false;
    while (Date.now() < deadline) {
      pub.publish(topic).message(Buffer.from(sent)).submit();
      try {
        if (sub.subscribe(received, zlink.RecvFlags.DontWait)) {
          hasReceived = true;
          break;
        }
      } catch (error) {
        if (!(error instanceof zlink.RecvError && error.result === zlink.RecvResult.NoData)) {
          throw error;
        }
      }
      await new Promise((resolve) => setTimeout(resolve, 25));
    }
    assert.equal(hasReceived, true);
    try {
      const recv = received.parts[0].data().toString();
      assert.equal(received.topic, topic);
      assert.equal(recv, sent);
      console.log(`[pubsub/recv] publish: "${topic}/${sent}" \u2192 subscribe: "${topic}/${recv}"`);
    } finally {
      received.close();
    }
  } finally {
    sub.close();
    pub.close();
    ctx.close();
  }
// --8<-- [end:doc]
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: bindings/python/samples/pubsub_recv_sample.py
import zlink
from sample_support import tcp_endpoint, wait_connected


def main():
# --8<-- [start:doc]
    _, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_xpub_socket(ctx) as publisher:
            with zlink.create_sub_socket(ctx) as subscriber:
                with publisher.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as publisher_monitor:
                    with subscriber.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as subscriber_monitor:
                        publisher.bind(endpoint)
                        subscriber.connect(endpoint)
                        subscriber.set_subscription(b"prices")
                        wait_connected(publisher_monitor, subscriber_monitor)

                event = zlink.create_subscription_event()
                if not publisher.receive_subscription_event_into(event):
                    raise AssertionError("missing subscription event")
                if not event.subscribed or event.topic != "prices":
                    raise AssertionError("unexpected subscription event")
                publisher.publish(b"prices").message(b"101.25").submit()
                received = zlink.create_topic_message()
                assert subscriber.subscribe_into(received)
                with received:
                    if received.topic != "prices":
                        raise AssertionError(f"unexpected pubsub topic: {received.topic!r}")
                    if received.to_bytes_list() != [b"101.25"]:
                        raise AssertionError("unexpected pubsub payload")
                print('[pubsub/recv] publish: "prices/101.25" → subscribe: "prices/101.25"')
# --8<-- [end:doc]


if __name__ == "__main__":
    main()
